src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeGenericConstraints.cs

```diff
@@ -0,0 +1,167 @@
+    ///   TypeParameter names should not change in strict mode, or when specified
+    ///   strict mode: no changes at all
+    ///   sealed types and non-virtual methods: no additions, removals permitted
+    ///   non-sealed types and virtual methods: no changes
+    /// </summary>
+    public class CannotChangeGenericConstraints : IRule
+    {
+        private readonly IRuleSettings _settings;
+
+        public CannotChangeGenericConstraints(IRuleSettings settings, IRuleRegistrationContext context)
+        {
+            _settings = settings;
+            context.RegisterOnMemberSymbolAction(RunOnMemberSymbol);
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol? left,
+            ITypeSymbol? right,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is not INamedTypeSymbol leftType || right is not INamedTypeSymbol rightType)
+            {
+                return;
+            }
+
+            var leftTypeParameters = leftType.TypeParameters;
+            var rightTypeParameters = rightType.TypeParameters;
+
+            // can remove constraints on sealed classes since no code should observe broader set of type parameters
+            bool permitConstraintRemoval = !_settings.StrictMode && leftType.IsSealed;
+
+            CompareTypeParameters(leftTypeParameters, rightTypeParameters, left, permitConstraintRemoval, leftMetadata, rightMetadata, differences);
+        }
+
+        private void RunOnMemberSymbol(
+            ISymbol? left,
+            ISymbol? right,
+            ITypeSymbol leftContainingType,
+            ITypeSymbol rightContainingType,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is not IMethodSymbol leftMethod || right is not IMethodSymbol rightMethod)
+            {
+                return;
+            }
+
+            var leftTypeParameters = leftMethod.TypeParameters;
+            var rightTypeParameters = rightMethod.TypeParameters;
+
+            bool permitConstraintRemoval = !_settings.StrictMode && !leftMethod.IsVirtual;
+
+            CompareTypeParameters(leftTypeParameters, rightTypeParameters, left, permitConstraintRemoval, leftMetadata, rightMetadata, differences);
+        }
+
+        private void CompareTypeParameters(
+            ImmutableArray<ITypeParameterSymbol> leftTypeParameters,
+            ImmutableArray<ITypeParameterSymbol> rightTypeParameters,
+            ISymbol left,
+            bool permitConstraintRemoval,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+
+            Debug.Assert(leftTypeParameters.Length == rightTypeParameters.Length);
```

Comment:
Is it guaranteed that the number of type parameters match?

Reply:
Yes, this is ensured by the mapper which aligns the type or method.

Reply:
Here's where I copied this from: https://github.com/dotnet/sdk/blob/d0f2b99de74845e15fa6dd1652566f47d8b1051b/src/Compatibility/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeParameterName.cs#L31